Utils/src/BTaggingEffAnalyzer.cc

```diff
@@ -0,0 +1,233 @@
+  ptNBins(iConfig.getParameter<int>("PtNBins")),
+  ptMin(iConfig.getParameter<double>("PtMin")),
+  ptMax(iConfig.getParameter<double>("PtMax")),
+  etaNBins(iConfig.getParameter<int>("EtaNBins")),
+  etaMin(iConfig.getParameter<double>("EtaMin")),
+  etaMax(iConfig.getParameter<double>("EtaMax"))
+
+{
+   //now do what ever initialization is needed
+   h2_BTaggingEff_Denom_b    = fs->make<TH2D>("h2_BTaggingEff_Denom_b", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Denom_c    = fs->make<TH2D>("h2_BTaggingEff_Denom_c", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Denom_udsg = fs->make<TH2D>("h2_BTaggingEff_Denom_udsg", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Num_b    = fs->make<TH2D>("h2_BTaggingEff_Num_b", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Num_c    = fs->make<TH2D>("h2_BTaggingEff_Num_c", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+   h2_BTaggingEff_Num_udsg = fs->make<TH2D>("h2_BTaggingEff_Num_udsg", ";p_{T} [GeV];#eta", ptNBins, ptMin, ptMax, etaNBins, etaMin, etaMax);
+}
+
+BTaggingEffAnalyzer::~BTaggingEffAnalyzer()
+{
+ 
+   // do anything here that needs to be done at desctruction time
+   // (e.g. close files, deallocate resources etc.)
+
+}
+
+
+//
+// member functions
+//
+
+// ------------ method called for each event  ------------
+void
+BTaggingEffAnalyzer::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  edm::Handle<PatJetCollection> jets;
+  iEvent.getByLabel(jetsTag,jets);
+
+  //grab event selection variables:
+  edm::Handle<double> HT;
+  edm::Handle<double> MHT; 
+  edm::Handle<double>DeltaPhi1;
+  edm::Handle<double>DeltaPhi2;
+  edm::Handle<double>DeltaPhi3;
+  edm::Handle<double>DeltaPhi4;
+  edm::Handle<int> NJets;
+  edm::Handle<int> Leptons;
+  edm::Handle<int> isoPionTracks;
+  edm::Handle<int> isoMuonTracks;
+  edm::Handle<int> isoElectronTracks;
+  edm::Handle<bool> JetID; 
+
+  iEvent.getByLabel("HT", HT);
+  iEvent.getByLabel("MHT","Pt", MHT);
+  iEvent.getByLabel("DeltaPhi", "DeltaPhi1", DeltaPhi1);
+  iEvent.getByLabel("DeltaPhi","DeltaPhi2", DeltaPhi2);
+  iEvent.getByLabel("DeltaPhi","DeltaPhi3", DeltaPhi3);
+//  iEvent.getByLabel("DeltaPhi","DeltaPhi4", DeltaPhi4);
+
+  iEvent.getByLabel("NJets", NJets);
+  iEvent.getByLabel("LeptonsNew", Leptons);
+  iEvent.getByLabel("IsolatedPionTracksVeto","isoTracks", isoPionTracks);
+  iEvent.getByLabel("IsolatedMuonTracksVeto","isoTracks", isoMuonTracks);
+  iEvent.getByLabel("IsolatedElectronTracksVeto","isoTracks", isoElectronTracks);
+  iEvent.getByLabel("GoodJets", JetID);
+  bool passEvSel=false;
+  if(*HT>500 && *MHT>200 && *DeltaPhi1>0.5 && *DeltaPhi2>0.5 && *DeltaPhi3>0.3 && *NJets>=4 && *Leptons==0 && *isoPionTracks==0 && *isoMuonTracks==0 && *isoElectronTracks==0 && *JetID)passEvSel=true;
```

Comment:
I was thinking you would run the efficiency plots whenever you run over the full signal sample for the first time. Then the histograms are stored in an original set of trees and you don’t run this module again. 
-Rishi
> On Oct 5, 2015, at 5:23 PM, Kevin Pedro <notifications@github.com> wrote:
> 
> In Utils/src/BTaggingEffAnalyzer.cc <https://github.com/TreeMaker/TreeMaker/pull/68#discussion_r41156799>:
> 
> > +
> > +  iEvent.getByLabel("HT", HT);
> > +  iEvent.getByLabel("MHT","Pt", MHT);
> > +  iEvent.getByLabel("DeltaPhi", "DeltaPhi1", DeltaPhi1);
> > +  iEvent.getByLabel("DeltaPhi","DeltaPhi2", DeltaPhi2);
> > +  iEvent.getByLabel("DeltaPhi","DeltaPhi3", DeltaPhi3);
> > +//  iEvent.getByLabel("DeltaPhi","DeltaPhi4", DeltaPhi4);
> > +
> > +  iEvent.getByLabel("NJets", NJets);
> > +  iEvent.getByLabel("LeptonsNew", Leptons);
> > +  iEvent.getByLabel("IsolatedPionTracksVeto","isoTracks", isoPionTracks);
> > +  iEvent.getByLabel("IsolatedMuonTracksVeto","isoTracks", isoMuonTracks);
> > +  iEvent.getByLabel("IsolatedElectronTracksVeto","isoTracks", isoElectronTracks);
> > +  iEvent.getByLabel("GoodJets", JetID);
> > +  bool passEvSel=false;
> > +  if(*HT>500 && *MHT>200 && *DeltaPhi1>0.5 && *DeltaPhi2>0.5 && *DeltaPhi3>0.3 && *NJets>=4 && *Leptons==0 && *isoPionTracks==0 && *isoMuonTracks==0 && *isoElectronTracks==0 && *JetID)passEvSel=true;
> I'm not really thrilled about implementing the whole RA2/b-specific signal region selection in a TreeMaker module. Is it absolutely necessary for this workflow, or could the b-tagging efficiency analyzer be kept separate from TreeMaker somehow?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/TreeMaker/TreeMaker/pull/68/files#r41156799>.
>